typing/parmatch.ml

```diff
@@ -42,7 +42,158 @@ let rec omegas i =
 
 let omega_list l = List.map (fun _ -> omega) l
 
-let zero = make_pat (Tpat_constant (Const_int 0)) Ctype.none Env.empty
+module Pattern_head : sig
+  type desc =
+    | Any
+    | Construct of constructor_description
+    | Constant of constant
+    | Tuple of int
+    | Record of label_description list
+    | Variant of
+        { tag: label; has_arg: bool;
+          cstr_row: row_desc ref;
+          type_row : unit -> row_desc; }
+          (* the row of the type may evolve if [close_variant] is called,
+             hence the (unit -> ...) delay *)
+    | Array of int
+    | Lazy
+
+  type t
```

Comment:
Why an abstract type + accessors instead of `type t = private { ... }`? This way around causes a lot of `| p -> match Pattern_head.desc p with ...` which could be `| {desc = ...; _} -> ...`

Reply:
Personal preference I guess.
Also, I don't it really causes many `| p -> match ..`, it does cause a fair number of 
```ocaml
| ((head, _), _) :: rows ->
  match Pattern_head.desc head with
  | Any -> ..
  | Tuple _ | Lazy -> ...
```
which I find more readable than
```ocaml
| (({ desc = Any ; _ }, _), _) :: rows ->  ...
| (({ desc = Tuple _ | Lazy ; _ }, _), _) :: rows ->  ...
```

Reply:
OK, makes sense!